buttons/messages/views: Add support for copying code snippets.

yuktasarode opened the pull request:
<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->

**What does this PR do?**  <!-- Overall description goes here -->
This PR aims to add support for copying code snippets and is based off on the work done by @kingjuno and @neiljp in the PR #1134. This PR mainly fixes the tests and linting issues and bring the PR to a merge-able state. 

<!-- If fixing a filed bug or new feature, add 'Fixes #<issue>' or 'Partial fix for #<issue>' -->

<!-- Add a link to a discussion on chat.zulip.org, if relevant -->

**Tested?** <!-- Fine to leave some of these unchecked if this is a draft/work-in-progress -->
- [x] Manually
- [x] Existing tests (adapted, if necessary)
- [ ] New tests added (for any new behavior)
- [x] Passed linting & tests (each commit)
<!-- Code must pass CI (GitHub Actions) before merging - look for the green tick! -->

This may fix the core elements of #1123.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1646,12 +1665,68 @@ def create_link_buttons(
+                #     display_code[-1] = tuple(display_code[-1])
+            code = f"{str(index+1)}: {language}\n"
+            button = CodeSnippetButton([("pygments:w", code)] + display_code)
+            urwid.connect_signal(button, "click", self.copy_to_clipboard, copy_code)
+            display_attr = None if index % 2 else "popup_contrast"
```
It would be cleaner if we could have this kind of behavior, and that in a few other methods below, defined in the CodeSnippetButton itself instead, like we do with other buttons - I can't see a reason not to do this.

Note how the create_link_button function is simpler compared to this one.

The CodeSnippetButton does very little that's unique right now - it's very close to being a very simple Button.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -1646,12 +1665,68 @@ def create_link_buttons(
+                code_snippet_width = max(
+                    code_snippet_width, len(max(display_code, key=len))
+                )
+                # if language == "":
+                #     display_code[-1] = list(display_code[-1])
+                #     display_code[-1][-1] = (
+                #         display_code[-1][-1][:-1]
+                #         if display_code[-1][-1][-1] == "\n"
+                #         else display_code
+                #     )
+                #     display_code[-1] = tuple(display_code[-1])
```
We should remove commented code from before; you can split this into a later commit, in case we want to refer to it for a later feature, or remove it and we can refer to the original PR instead.

yuktasarode replied:
I have deleted the commented code.

neiljp commented on zulipterminal/ui_tools/messages.py:
```diff
@@ -551,8 +560,16 @@ def soup2markup(
                     else:
                         markup.append(("pygments:w", code_text))
+                        code_snippet.append(("pygments:w", code_text))
+                try:
+                    code_language = element.attrs["data-code-language"]
+                except KeyError:
+                    code_language = None
```
We can simplify this, since attrs is essentially a dict, using .get.

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -647,6 +647,40 @@ def handle_narrow_link(self) -> None:
                 self.controller.exit_popup()
 
 
+class CodeSnippetButton(urwid.Text):
```
As mentioned elsewhere, this works, but it would be good to migrate to a button with the specific functionality we want.

yuktasarode replied:
I did trying inherting urwid.Button and Moving the behavior into the CodeSnippetButton but I could only move the method get_code_from_snippet() from views.py to buttons.py, since all other methods require the instance of controller to be taken into account. Kindly let me know anyother way to do this.

neiljp replied:
Right now, all the basic Button functionality currently in `CodeSnippetButton` except for `get_code_from_snippet` is as per the default in `urwid.Button` (you can compare yourself, to `wimp.py` in urwid).

Look at the initializer for `MessageLinkButton` - that is how we can set up a controller via the button. We may also want to store the code content (snippet information) in there, for example.

Right now you're using the class name to call a method in a class, and explicitly passing the `self` value. Normally that'd be a bad idea and often cause an error, but you're not using `self` in `get_code_from_snippet` here! Instead, we should be creating `CodeSnippetButton`s with the relevant data, then using their properties as we need them.

I understand you've inherited this PR, but this (and the UI, using urwid) is using OOP in Python, so you may need to learn more about this if you've not worked on this before.

neiljp reviewed (commented):
@yuktasarode Thanks for tidying this up further! :+1:

It's good to see this passing linting and tests :) It's also working well still, though I've not examined the functionality in great detail yet.

I've left some comments inline, but you may want to first rebase against `main`, since I recently merged a PR regarding how the 'Action' keys are shown and grouped, which affects this UI.

My note in the other PR (https://github.com/zulip/zulip-terminal/pull/1134#issuecomment-1472479750) mention a few other changes; I'm not sure if you're already looking at those aspects, but I've left some details below.

I'd suggest the next step to make this easier to work on and read would be to split this into multiple commits. You'll want to confirm how the code links together yourself, but my take would be that we're generating additional data via the message, which will need to go into a button, both of which are combined into the end result - so 3 commits might be a good starting point.

Moving the behavior into the CodeSnippetButton may be another good step, or at least inheriting via `urwid.Button` like the MessageLinkButton does in the first instance, since I believe it should achieve most of what the custom button does already?

My note in the other PR (https://github.com/zulip/zulip-terminal/pull/1134#issuecomment-1472479750) also notes regarding adding tests for each commit/change; if you wish to start on that, that's fine, but I expect it'll be a lot easier to keep track of when the commit size is smaller, and you can separate which code change/addition is related to which test change/addition.

neiljp commented:
@yuktasarode Is this ready for review? If so, please change the labels as per my message in the GSoC 2023 topic in zulip-terminal (ie. remove needing update and add needs review).

yuktasarode commented:
@zulipbot add "PR needs review"

yuktasarode commented:
@zulipbot remove "PR awaiting update"

neiljp commented:
@yuktasarode Could you please rebase this after the recent changes we had in `main`, and further explore how you might simplify and initialize the snippet button like we do with the `MessageLinkButton`? (given my comment above)

You're welcome to take up another issue also, of course :+1:

yuktasarode committed:
buttons: Added CodeSnippetButton class.

Made the button inhert urwid.button and added supporting methods for creating the code snippet button.

yuktasarode committed:
core/views: Populate code snippet button with data.

Inherited urwid.button class in create_code_snippet_buttons method.
Get the code snippets and copy it on button press.
Tests added for create_code_snippet_buttons.

yuktasarode committed:
messages/views: Add markup to code snippets.

Tests updated.

neiljp reviewed (commented):
@yuktasarode You didn't note this for review, but I saw you had pushed. This looks much closer than before :tada:
